File: Drum/sticking-engine/sticking_engine/notation.py
"""
notation.py — convenience export helpers (not a numbered rule).

Turns a FinalValidatedPerformance into human-readable sticking notation
(R/L per hand event) and a simple JSON-serializable dict, for quick
inspection, debugging, or feeding into another tool.
"""

from __future__ import annotations
import json
from typing import Dict, Any
from .datamodel import FinalValidatedPerformance, Limb


def to_sticking_line(performance: FinalValidatedPerformance, instrument_filter=None) -> str:
    """Return a single-line R/L/RF/LF sticking string in time order."""
    events = sorted(performance.events, key=lambda e: e.time_seconds)
    if instrument_filter:
        events = [e for e in events if e.instrument in instrument_filter]
    return " ".join(e.limb.value for e in events)


def to_dict(performance: FinalValidatedPerformance) -> Dict[str, Any]:
    return {
        "duration_s": performance.duration_s,
        "seed": performance.seed,
        "engine_version": performance.engine_version,
        "approved": performance.validation.approved,
        "validation_issues": [
            {"severity": i.severity.value, "code": i.code, "message": i.message, "event_id": i.event_id}
            for i in performance.validation.issues
        ],
        "events": [
            {
                "event_id": e.event_id,
                "time_seconds": round(e.time_seconds, 4),
                "limb": e.limb.value,
                "instrument": e.instrument.value,
                "stroke_type": e.stroke_type.value,
                "velocity": e.velocity,
                "microtiming_offset_ms": round(e.microtiming_offset_ms, 2),
                "dynamic_level": round(e.dynamic_level, 3),
            }
            for e in sorted(performance.events, key=lambda ev: ev.time_seconds)
        ],
    }


def to_json(performance: FinalValidatedPerformance, path: str) -> None:
    with open(path, "w") as f:
        json.dump(to_dict(performance), f, indent=2)


def print_table(performance: FinalValidatedPerformance, limit: int = 40) -> None:
    events = sorted(performance.events, key=lambda e: e.time_seconds)
    print(f"{'time':>8} {'limb':>4} {'instrument':<16} {'stroke':<10} {'vel':>4}")
    for e in events[:limit]:
        print(f"{e.time_seconds:8.3f} {e.limb.value:>4} {e.instrument.value:<16} "
              f"{e.stroke_type.value:<10} {e.velocity:4d}")
    if len(events) > limit:
        print(f"... ({len(events) - limit} more events)")

File: Drum/sticking-engine-ts/src/notation.ts
/**
 * notation.ts — text and JSON views of a solved performance.
 *
 * A helper, not a numbered rule. It reads `FinalValidatedPerformance`
 * and writes; it never decides anything.
 */

import type { FinalValidatedPerformance } from './datamodel.js';

/** The same columns the Python's `print_table` produces, as lines. */
export function performanceTable(performance: FinalValidatedPerformance): string[] {
  const lines = ['    time limb instrument       stroke      vel', '  ' + '-'.repeat(45)];
  for (const ev of performance.events) {
    lines.push(
      `${ev.timeSeconds.toFixed(3).padStart(8)} ${ev.limb.padStart(4)} ` +
        `${ev.instrument.padEnd(16)}${ev.strokeType.padEnd(12)}${String(ev.velocity).padStart(3)}`,
    );
  }
  lines.push('');
  lines.push(
    `Approved: ${performance.validation.approved}  ` +
      `Issues: ${performance.validation.issues.length}  ` +
      `Duration: ${performance.durationS.toFixed(2)}s`,
  );
  return lines;
}

/** A plain object ready for `JSON.stringify` — the port's parity output. */
export function toPlainObject(performance: FinalValidatedPerformance): unknown {
  return {
    seed: performance.seed,
    engine_version: performance.engineVersion,
    duration_s: performance.durationS,
    approved: performance.validation.approved,
    issues: performance.validation.issues.map((i) => ({
      severity: i.severity,
      code: i.code,
      event_id: i.eventId ?? null,
    })),
    events: performance.events.map((e) => ({
      event_id: e.eventId,
      source_id: e.sourceId,
      time_seconds: e.timeSeconds,
      limb: e.limb,
      instrument: e.instrument,
      stroke_type: e.strokeType,
      velocity: e.velocity,
      microtiming_offset_ms: e.microtimingOffsetMs,
      dynamic_level: e.dynamicLevel,
    })),
  };
}
